app/src/main/java/fi/bitrite/android/ws/model/Message.java

```diff
@@ -52,19 +54,31 @@ public Message cloneForIsPushed(boolean isPushed) {
     }
 
     /**
-     * Removes the <p>...</p>\r\n surrounding all messages sent from the web interface.
+     * Removes the "\r\n" between two "<p>...</p>"-blocks.
+     *
+     * WS creates a new "<p>...</p>" block whenever the user presses <enter> to end the current line
+     * of a message in the web form.
+     * If the user presses <shift>-<enter> a <br/> is added to the current "<p>...</p>"-block.
      */
-    private static String stripRawBody(String rawBody) {
-        if (rawBody.startsWith("<p>") && rawBody.endsWith("</p>\r\n")) {
-            rawBody = rawBody.substring(3, rawBody.length()-6);
+    private static String stripBody(String rawBody) {
+        if (!rawBody.startsWith("<p>")) {
+            // This message was generated by email reply. It is not enclosed by <p>...</p> and uses
+            // '\n' for newlines.
+            rawBody = "<p>"
+                      + rawBody.replace("\r", "")
+                              .replace("\n", "<br>")
+                      + "</p>";
         }
-        return rawBody;
+        return rawBody.replace("\r\n", "");
     }
+
     /**
-     * Replaces \n by <br/> and then returns the html-parsed body.
+     * Returns the html-parsed body and removes the trailing newlines caused by the last paragraph.
      */
-    private static Spanned parseBody(String rawBody) {
-        rawBody = rawBody.replace("\n", "<br/>");
-        return Html.fromHtml(rawBody);
+    private static CharSequence parseBody(String strippedRawBody) {
+        Spanned body = Html.fromHtml(strippedRawBody);
+        int i = body.length()-1;
+        for (; i >= 0 && Character.isWhitespace(body.charAt(i)); --i);
```

Comment:
What is happening here?! Regarding code readability, there is much to be desired here.

Reply:
The way I understand the function, couldn’t we just replace it with the following?

```suggestion
private static CharSequence parseBody(String strippedRawBody) {
    return Html.fromHtml(strippedRawBody.trim());
}
```

Reply:
Added a comment that should explain the issue.
`Html.fromHtml("<p>xxx</p>") == "xxx\n\n"`. Here, we strip the last two newlines.